src/uproot/models/RNTuple.py

```diff
@@ -218,26 +217,56 @@ def page_list_envelopes(self):
 
         return self.column_innerlist_dict
 
-    def col_form(self, field_id):
-        for cr in self.header.column_records:
+    def base_col_form(self, cr, col_id, parameters=None):
+        form_key = f"column-{col_id}"
+        if cr.type == uproot.const.rt_role_union:  # switch (UnionForm)
+            return form_key
+        elif cr.type > uproot.const.rt_role_struct:  # data column
+            return ak._v2.forms.NumpyForm(
+                _rntuple_col_types[cr.type], form_key=form_key, parameters=parameters
+            )
+        else:  # offset index column
+            return form_key
+
+    def col_form(self, field_id, type_name=None):
+        # FIXME remove this ugly logic
+        rel_crs = []
+        rel_crs_idxs = []
+        for i, cr in enumerate(self.header.column_records):
             if cr.field_id == field_id:
-                form_key = f"field-{cr.field_id}"
-                if cr.type > 2:  # data column
-                    return ak._v2.forms.NumpyForm(
-                        _rntuple_col_types[cr.type], form_key=form_key
-                    )
-                else:  # offset index column
-                    return form_key
+                rel_crs.append(cr)
+                rel_crs_idxs.append(i)
+            if cr.field_id > field_id:
+                break
+        if len(rel_crs) == 1:  # base case
+            return self.base_col_form(
+                rel_crs[0], rel_crs_idxs[0]
+            )
+        elif type_name == "std::string":  # string field splits->2 in col records
+            assert len(rel_crs_idxs) == 2
+            cr_char = rel_crs[-1]
+            assert cr_char.type == 5  # char
+            inner = self.base_col_form(
+                cr_char, rel_crs_idxs[-1], parameters={"__array__": "char"}
+            )
+            form_key = f"column-{rel_crs_idxs[0]}"
+            return ak._v2.forms.ListOffsetForm(
+                "u32", inner, form_key=form_key, parameters={"__array__": "string"}
+            )
+        else:
+            raise (RuntimeError(f"Missing special case: {field_id}"))
 
     def field_form(self, this_id, seen):
         frs = self.header.field_records
         fr = frs[this_id]
         seen.append(this_id)
-        if fr.struct_role == 0:
+        sr = fr.struct_role
+        if sr == 0:
             # base case of recursive
-            # these two roles have exactly one column belong to them
-            return self.col_form(this_id)
-        elif fr.struct_role == 1:
+            # n.b. the split may be in column
+            return self.col_form(this_id, fr.type_name)
+        elif sr == 1:
+            # mother of a collection, i.e. vector (meaning field is jagged)
```

Comment:
More instances of named constants ("`0`" and "`1`").

------------------

But also, why abbreviate names like `self.header.field_records` → `frs`? It makes me have to go look them up. From the perspective of writing code, using long, fully qualified names feels redundant and you want to adopt a shorthand. But from the perspective of reading code, especially when hunting a bug or something where I'm only interested in a small local part of the codebase, answering questions like "what does "`frs`" mean?" is an extra burden.

I prefer fully qualified names, even if they're pretty long, to prioritize the needs of code-readers (over code-writers). I'm particularly aggravated when reading Python code that has a lot of `from some.library import SomeClass` at the top of the file, so when I encounter `SomeClass`, I have to go up to the top of the file to find out what it is and where it's defined. Maybe code analysis tools help with that, but those tools will get confused if there's too much/too complex indirection. I had to be won over to the idea of `import numpy as np`, but became okay with that when I saw how standard that particular substitution was. It's legal to `import numpy as qz`, but that would be terrible.

The downside of using fully qualified names almost everywhere is that code formatters limit the line length. I haven't entirely figured out how to reconcile the value of fully qualified names with how much black obfuscates a line when it's really long. So you don't have to use

```python
self.header.field_records[this_id]
```

everywhere you currently use

```python
fr
```

and get those deeply nested black-formattings. But maybe the substitutions don't have to be 2 and 3 letters long:

```python
field_records = self.header.field_records
field_record = field_records[this_id]
```

(and use them consistently in any function that uses it, kind of like how "`np`" always means "`numpy`")?